### DISCUSSION TITLE
Lab: Repeated Operations I

### DISCUSSION THREAD

**Discussion by mbcann01** [brad-cannell/intro.r.public#25] [Labs]
Please feel free to post questions about this lab below. You are also welcome to leave comments below about the usefulness of this lab. Which parts were helpful? What could I do to improve it? What is still unclear?

**Comment by OJONESD** [comment]
Hello Dr. Cannell,
I had a question about Task 7 "Please recode all Not Sure responses in the class_survey data frame to missing."
I looked at the reference code for the lab and I still get an error (pictured below). Did I miss a step or is there another way to do this task?
I got correct column for the question 14 but my numbers didn't match
Thank you for your help,
Olivia Jones
<img width="386" alt="image" src="https://github.com/user-attachments/assets/3c0022bd-2241-4151-b0f1-d5cfd979b1c1">

**Reply by mbcann01** [reply]
Hi @OJONESD ,

This is a great question. It looks like this error is occurring in the `if_else()` line of code. Specifically, when the `across()` function passes the `required` column to the `if_else()` line of code. That line of code telling R:

- If the value in the current row of the `required` column is 9 or 99, then set the value of `required` to NA_real_ -- a missing value in a numeric vector. This is also called the TRUE value -- What R should do if the if_else condition is TRUE.
- If the value in the current row of the `required` column is _not_ 9 or 99,  then set the value of `required` to the same value it currently has. This is also the FALSE value -- What R should do if the if_else condition is FALSE.
- For safety reasons, [the `if_else()` function doesn't allow us to combine different vector types together](https://dplyr.tidyverse.org/reference/if_else.html). In other words, the TRUE value and the FALSE value must be of the same type (e.g., numeric, character, logical, etc.).
- The error in the screenshot above says, "can't combine `true` <double> and `false` <character>." In other words, we are trying to combine a numeric vector and a character vector. And it even tells us which is which. The TRUE value is numeric (double) and the false value is character. 
- We know that we the true value should be numeric because we explicitly set it to be numeric in the true condition (i.e., `NA_real_`). See bullet 1 above. 
- R is telling us that the false value is character. What is the false value? In bullet two above, we already determined that the value R should use if the if_else condition was false was the current value of `required`. Therefore, it seems like `required` is currently a character vector. 
- At this point, we have two options to get the `if_else()` code to work. We can change the true value to character or the false value to numeric. Remember, they have to match. But, because we want the final result to be a numeric vector, we should change the false value to numeric. In fact, we should have already done so earlier in the code. But perhaps it didn't work for some reason.

I think your next steps are to check and see if `required` is currently a character vector. If so, then coerce it to be numeric. I hope that helps!

**Comment by OJONESD** [comment]
I just wanted to comment the repeated operations chapters are very helpful, I definitely plan on getting more experience with it. While working on my Intermediate Biostatistics final project in Stata I kept thinking how it would be better in R because I was having to repeat very similar lines of code over and over. Thank you again for putting together the lab.
Olivia Jones

**Reply by mbcann01** [reply]
Thank you, @OJONESD! I'm really glad you found it helpful!
